timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/BatchingPaxosLearnerFactory.java

```diff
@@ -0,0 +1,146 @@
+            DisruptorAutobatcher<WithSeq<Client>, PaxosResponses<PaxosContainer<Optional<PaxosValue>>>> getLearnedValues,
+            DisruptorAutobatcher<WithSeq<Client>, PaxosResponses<PaxosUpdate>> getLearnedValuesSince) {
+        this.learn = learn;
+        this.getLearnedValues = getLearnedValues;
+        this.getLearnedValuesSince = getLearnedValuesSince;
+    }
+
+    public static BatchingPaxosLearnerFactory create(
+            List<BatchPaxosLearner> learners,
+            ExecutorService executor,
+            int quorumSize) {
+        DisruptorAutobatcher<Map.Entry<Client, PaxosValue>, PaxosResponses<PaxosResponse>> learn =
+                Autobatchers.coalescing(wrap(learners, executor, quorumSize, LearnCoalescingConsumer::new))
+                        .safeLoggablePurpose("batch-paxos-learner.learn")
+                        .build();
+
+        DisruptorAutobatcher<WithSeq<Client>, PaxosResponses<PaxosContainer<Optional<PaxosValue>>>> learnedValues =
+                Autobatchers.coalescing(wrap(learners, executor, quorumSize, LearnedValuesCoalescingFunction::new))
+                        .safeLoggablePurpose("batch-paxos-learner.learned-values")
+                        .build();
+
+        DisruptorAutobatcher<WithSeq<Client>, PaxosResponses<PaxosUpdate>> learnedValuesSince =
+                Autobatchers.coalescing(wrap(learners, executor, quorumSize, LearnedValuesSinceCoalescingFunction::new))
+                .safeLoggablePurpose("batch-paxos-learner.learned-values-since")
+                .build();
+
+        return new BatchingPaxosLearnerFactory(
+                learn,
+                learnedValues,
+                learnedValuesSince);
+    }
+
+    public PaxosLearnerNetworkClient paxosLearnerForClient(Client client) {
+        return new AutobatchingPaxosLearnerNetworkClient(client);
+    }
+
+    private final class AutobatchingPaxosLearnerNetworkClient implements PaxosLearnerNetworkClient {
+        private final Client client;
+
+        private AutobatchingPaxosLearnerNetworkClient(Client client) {
+            this.client = client;
+        }
+
+        @Override
+        public void learn(long seq, PaxosValue value) {
+
+        }
+
+        @Override
+        public <T extends PaxosResponse> PaxosResponses<T> getLearnedValue(
+                long seq,
+                Function<Optional<PaxosValue>, T> mapper) {
+            try {
+                return getLearnedValues.apply(WithSeq.of(seq, client)).get().map(c -> mapper.apply(c.get()));
+            } catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                }
+                throw new RuntimeException(cause);
+            } catch (InterruptedException e) {
+                // TODO(fdesouza): handle
+                throw new RuntimeException(e);
```

Comment:
I don't think the 3rd option is valuable. If I'm interrupted, => I didn't get a response, so unless I have some sort of error type, there is no difference between interrupted and just some random network error.

Come to think of it, 1. isn't really viable, because every where we used to call `PaxosLearner` and `PaxosAcceptor` (`Paxos*NetworkClient` now), would have to deal and understand the interrupts, but I don't believe it's useful to do so. 

I think I'll just go with option 2.

Reply:
I have some code that centralises exception handling in #4072, I will unify them when these merge if that's okay?

Reply:
Yeah, that's fine. With the options I'm usually and in this case as well in favour of 2.